game_modules/game_engine.py

```diff
@@ -0,0 +1,822 @@
+    loopin aikaisista tapahtumista kuten nappien painalluksista.
+    """
+    def game_loop(self):
+        # Peli looppi
+        self.apple.newApple(self.gridSize)
+        # Musiikki
+        pygame.mixer.music.play(-1)
+        # Jos vaikeustaso on "Hard", generoidaan esteitä
+        if (self.difficulty.name == "Hard"):
+            for i in range(random.randint(7, 15)):
+                self.obstacles.append(Obstacle(self.windowWidth / self.gridSize, self.windowHeight / self.gridSize))
+                self.obstacles[i].newObstacle(self.gridSize)
+
+        while self.running:
+            # Tapahtuma looppi
+            for event in pygame.event.get():
+                # Ensimmäinen if lause käsittelee pelistä poistumisen
+                if event.type == QUIT:
+                    pygame.quit()
+                    quit()
+
+                elif event.type == KEYDOWN:
+                    now = pygame.time.get_ticks()
+                    now2 = pygame.time.get_ticks()
+                    #Pelin pysäyttävät näppäimet
+                    if event.key == K_ESCAPE:
+                        self.pause = not self.pause
+                    if event.key == K_SPACE:
+                        self.pause = not self.pause
+                    if event.key == K_p:
+                        self.pause = not self.pause
+
+                    if event.key == K_m:
+                        self.sound = not self.sound
+
+                    # Cooldowneilla estetään nappien spämmäys
+                    # Pelaaja 1
+                    if event.key == K_RIGHT and now - self.last >= self.cooldown:
+                        self.last = now
+                        self.snake.moveRight()
+
+                    if event.key == K_LEFT and now - self.last >= self.cooldown:
+                        self.last = now
+                        self.snake.moveLeft()
+
+                    if event.key == K_UP and now - self.last >= self.cooldown:
+                        self.last = now
+                        self.snake.moveUp()
+
+                    if event.key == K_DOWN and now - self.last >= self.cooldown:
+                        self.last = now
+                        self.snake.moveDown()
+                    # Pelaaja 2
+                    if event.key == K_d and now2 - self.last2 >= self.cooldown:
+                        self.last2 = now2
+                        self.other_snake.moveRight()
+
+                    if event.key == K_a and now2 - self.last2 >= self.cooldown:
+                        self.last2 = now2
+                        self.other_snake.moveLeft()
+
+                    if event.key == K_w and now2 - self.last2 >= self.cooldown:
+                        self.last2 = now2
+                        self.other_snake.moveUp()
+
+                    if event.key == K_s and now2 - self.last2 >= self.cooldown:
+                        self.last2 = now2
+                        self.other_snake.moveDown()
```

Comment:
needless repetition + fragility, these all do pretty much the same thing, we should loop through all keys and users and only implement this once in a general way